lib/builder/target.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Burrito.Builder.Target do
+  use TypedStruct
+
+  alias __MODULE__
+  alias Burrito.Builder.Log
+  alias Burrito.Util
+
+  @old_targets [:darwin, :win64, :linux, :linux_musl]
+
+  @required_key [:os, :cpu]
+
+  typedstruct do
+    field :alias, atom(), enforce: true
+    field :cpu, atom(), enforce: true
+    field :os, atom(), enforce: true
+    field :qualifiers, keyword(), enforce: true
+    field :otp_version, String.t(), enforce: true
+    field :debug?, boolean(), enforce: true
+  end
+
+  @spec init_target(keyword(), atom(), boolean()) :: Burrito.Builder.Target.t()
+  def init_target(target, target_alias, debug?) do
```

Comment:
I notice the logic for constructing and validating this struct is spread between this module and `Burrito.Builder`. It may be worth trying to encapsulate some of that logic into the constructor here.

This is untested pseudocode, but something along these lines may make the structure of the code easier to follow.

For demonstration purposes, I took a similar approach for handling the ERTS location to what I described here: https://github.com/burrito-elixir/burrito/pull/30/files#r770201823

This is a design that takes inspiration from `Mix.SCM`: https://github.com/elixir-lang/elixir/blob/79cd891eb8/lib/mix/lib/mix/scm.ex#L55

```elixir
defmodule Burrito.Builder.Target do
  ...

  @type legacy_definition :: atom()
  @type target_definition ::
          keyword()
          | legacy_definition()

  typedstruct enforce: true do
    field(:alias, atom())
    field(:cpu, atom())
    field(:os, atom())
    field(:debug?, boolean())
    field(:erts_manager, ERTSManager.t())
    field(:erts_opts, keyword())
  end

  @spec init_target(atom(), target_definition()) :: t()
  def init_target(target_alias, legacy_definition)
      when is_atom(legacy_definition) do
    definition =
      case legacy_definition do
        :darwin ->
          [os: :darwin, cpu: :x86_64]

        :win64 ->
          [os: :windows, cpu: :x86_64]

        :linux ->
          [os: :linux, cpu: :x86_64, libc: :glibc]

        :linux_musl ->
          [os: :linux, cpu: :x86_64, libc: :musl]

        other ->
          raise "Unrecognized legacy build target: #{other}"
      end

    Log.warning(
      :build,
      "You have specified an old-style build target: #{legacy_definition}\n\tPlease see the Burrito README for instructions on migrating to the new format!"
    )

    init_target(target_alias, definition)
  end

  def init_target(target_alias, definition) when is_list(definition) do
    {fields, rest} = Keyword.split(definition, [:os, :cpu, :debug?])

    {erts_manager, erts_opts} = get_erts_manager(rest)

    unless erts_manager do
      raise "Could not find an ERTS manager for build target #{inspect(target_alias)}"
    end

    fields =
      fields
      |> Keyword.put(:alias, target_alias)
      |> Keyword.put(:erts_manager, erts_manager)
      |> Keyword.put(:erts_opts, erts_opts)

    struct!(__MODULE__, fields)
  end

  defp get_erts_manager(opts) do
    # A list of modules implementing the ERTSManager behaviour,
    # which describes how to fetch an ERTS build
    erts_managers = ...

    Enum.find_value(erts_managers, {nil, opts}, fn manager ->
      # manager.accepts_options receives a keyword list of options and returns
      # an updated list of options if the manager can handle the given options,
      # and nil otherwise
      (new = manager.accepts_options(opts)) && {manager, new}
    end)
  end
end
```

Reply:
I was toying with relocating the other parts of the logic myself honestly, I'm leaning to moving it into the Target module!